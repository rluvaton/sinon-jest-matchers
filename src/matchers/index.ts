export { sinonToBeCalled } from './sinonToBeCalled';
export { sinonToBeCalledTimes } from './sinonToBeCalledTimes';
export { sinonToBeCalledWith } from './sinonToBeCalledWith';
export { sinonLastCalledWith } from './sinonLastCalledWith';
export { sinonLastReturnedWith } from './sinonLastReturnedWith';
export { sinonNthCalledWith } from './sinonNthCalledWith';
export { sinonToReturn } from './sinonToReturn';
export { sinonToReturnTimes } from './sinonToReturnTimes';
export { sinonToReturnWith } from './sinonToReturnWith';
