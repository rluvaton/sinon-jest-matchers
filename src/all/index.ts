import * as matchers from '../matchers';

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const jestExpect = global.expect;

if (jestExpect !== undefined) {
  jestExpect.extend(matchers);
} else {
  throw new Error(
    "Unable to find Jest's global expect. " +
      'Please check you have added sinon-jest-matchers correctly to your jest configuration. ' +
      'See https://github.com/rluvaton/sinon-jest-matchers#setup for help.',
  );
}
