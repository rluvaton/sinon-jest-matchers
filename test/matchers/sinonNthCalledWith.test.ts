import sinon from 'sinon';
import * as matchers from '../../src';
import * as Immutable from 'immutable';

expect.extend(matchers);

expect.extend({
  optionalFn(fn?: unknown) {
    const pass = fn === undefined || typeof fn === 'function';
    return { message: () => 'expect either a function or undefined', pass };
  },
});

describe('sinonNthCalledWith', () => {
  test('works only on stubs or sinon.spy', () => {
    const fn = function fn() {};

    expect(() => expect(fn).sinonNthCalledWith(1)).toThrowErrorMatchingSnapshot();
  });

  test('works when not called', () => {
    const fn = sinon.spy();
    expect(fn).not.sinonNthCalledWith(1, 'foo', 'bar');

    expect(() => expect(fn).sinonNthCalledWith(1, 'foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test('works with no arguments', () => {
    const fn = sinon.spy();
    fn();
    expect(fn).sinonNthCalledWith(1);
  });

  test("works with arguments that don't match", () => {
    const fn = sinon.spy();
    fn('foo', 'bar1');

    expect(fn).not.sinonNthCalledWith(1, 'foo', 'bar');

    expect(() => expect(fn).sinonNthCalledWith(1, 'foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test("works with arguments that don't match in number of arguments", () => {
    const fn = sinon.spy();
    fn('foo', 'bar', 'plop');

    expect(fn).not.sinonNthCalledWith(1, 'foo', 'bar');

    expect(() => expect(fn).sinonNthCalledWith(1, 'foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test("works with arguments that don't match with matchers", () => {
    const fn = sinon.spy();
    fn('foo', 'bar');

    expect(fn).not.sinonNthCalledWith(1, expect.any(String), expect.any(Number));

    expect(() =>
      expect(fn).sinonNthCalledWith(1, expect.any(String), expect.any(Number)),
    ).toThrowErrorMatchingSnapshot();
  });

  test("works with arguments that don't match with matchers even when argument is undefined", () => {
    const fn = sinon.spy();
    fn('foo', undefined);

    expect(fn).not.sinonNthCalledWith(1, 'foo', expect.any(String));

    expect(() => expect(fn).sinonNthCalledWith(1, 'foo', expect.any(String))).toThrowErrorMatchingSnapshot();
  });

  test('works with arguments that match', () => {
    const fn = sinon.spy();
    fn('foo', 'bar');

    expect(fn).sinonNthCalledWith(1, 'foo', 'bar');

    expect(() => expect(fn).not.sinonNthCalledWith(1, 'foo', 'bar')).toThrowErrorMatchingSnapshot();
  });

  test('works with arguments that match with matchers', () => {
    const fn = sinon.spy();
    fn('foo', 'bar');

    expect(fn).sinonNthCalledWith(1, expect.any(String), expect.any(String));

    expect(() =>
      expect(fn).not.sinonNthCalledWith(1, expect.any(String), expect.any(String)),
    ).toThrowErrorMatchingSnapshot();
  });

  test('works with trailing undefined arguments if requested by the match query', () => {
    const fn = sinon.spy();
    fn('foo', undefined);

    expect(fn).sinonNthCalledWith(1, 'foo', undefined);
    expect(() => expect(fn).not.sinonNthCalledWith(1, 'foo', undefined)).toThrowErrorMatchingSnapshot();
  });

  test('works with trailing undefined arguments when explicitly requested as optional by matcher', () => {
    // issue 12463
    const fn = sinon.spy();
    fn('foo', undefined);

    expect(fn).sinonNthCalledWith(1, 'foo', (expect as any).optionalFn());
    expect(() =>
      expect(fn).not.sinonNthCalledWith(1, 'foo', (expect as any).optionalFn()),
    ).toThrowErrorMatchingSnapshot();
  });

  test('works with Map', () => {
    const fn = sinon.spy();

    const m1 = new Map([
      [1, 2],
      [2, 1],
    ]);
    const m2 = new Map([
      [1, 2],
      [2, 1],
    ]);
    const m3 = new Map([
      ['a', 'b'],
      ['b', 'a'],
    ]);

    fn(m1);

    expect(fn).sinonNthCalledWith(1, m2);
    expect(fn).not.sinonNthCalledWith(1, m3);

    expect(() => expect(fn).not.sinonNthCalledWith(1, m2)).toThrowErrorMatchingSnapshot();
    expect(() => expect(fn).sinonNthCalledWith(1, m3)).toThrowErrorMatchingSnapshot();
  });

  test('works with Set', () => {
    const fn = sinon.spy();

    const s1 = new Set([1, 2]);
    const s2 = new Set([1, 2]);
    const s3 = new Set([3, 4]);

    fn(s1);

    expect(fn).sinonNthCalledWith(1, s2);
    expect(fn).not.sinonNthCalledWith(1, s3);

    expect(() => expect(fn).not.sinonNthCalledWith(1, s2)).toThrowErrorMatchingSnapshot();
    expect(() => expect(fn).sinonNthCalledWith(1, s3)).toThrowErrorMatchingSnapshot();
  });

  test('works with Immutable.js objects', () => {
    const fn = sinon.spy();
    const directlyCreated = Immutable.Map([['a', { b: 'c' }]]);
    const indirectlyCreated = Immutable.Map().set('a', { b: 'c' });
    fn(directlyCreated, indirectlyCreated);

    expect(fn).sinonNthCalledWith(1, indirectlyCreated, directlyCreated);

    expect(() =>
      expect(fn).not.sinonNthCalledWith(1, indirectlyCreated, directlyCreated),
    ).toThrowErrorMatchingSnapshot();
  });

  test('works with three calls', () => {
    const fn = sinon.spy();
    fn('foo1', 'bar');
    fn('foo', 'bar1');
    fn('foo', 'bar');

    expect(fn).sinonNthCalledWith(1, 'foo1', 'bar');
    expect(fn).sinonNthCalledWith(2, 'foo', 'bar1');
    expect(fn).sinonNthCalledWith(3, 'foo', 'bar');

    expect(() => {
      expect(fn).not.sinonNthCalledWith(1, 'foo1', 'bar');
    }).toThrowErrorMatchingSnapshot();
  });

  test('positive throw matcher error for n that is not positive integer', async () => {
    const fn = sinon.spy();
    fn('foo1', 'bar');

    expect(() => {
      expect(fn).sinonNthCalledWith(0, 'foo1', 'bar');
    }).toThrowErrorMatchingSnapshot();
  });

  test('positive throw matcher error for n that is not integer', async () => {
    const fn = sinon.spy();
    fn('foo1', 'bar');

    expect(() => {
      expect(fn).sinonNthCalledWith(0.1, 'foo1', 'bar');
    }).toThrowErrorMatchingSnapshot();
  });

  test('negative throw matcher error for n that is not integer', async () => {
    const fn = sinon.spy();
    fn('foo1', 'bar');

    expect(() => {
      expect(fn).not.sinonNthCalledWith(Infinity, 'foo1', 'bar');
    }).toThrowErrorMatchingSnapshot();
  });

  test('includes the custom mock name in the error message', () => {
    const fn = sinon.spy();
    fn.named('named-mock');
    fn('foo', 'bar');

    expect(fn).sinonNthCalledWith(1, 'foo', 'bar');

    expect(() => expect(fn).not.sinonNthCalledWith(1, 'foo', 'bar')).toThrowErrorMatchingSnapshot();
  });
});
