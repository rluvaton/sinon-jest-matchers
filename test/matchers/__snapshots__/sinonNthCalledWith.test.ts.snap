// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sinonNthCalledWith includes the custom mock name in the error message 1`] = `
"expect(named-mock).not.sinonNthCalledWith(n, ...expected)

n: 1
Expected: not \\"foo\\", \\"bar\\"

Number of calls: 1"
`;

exports[`sinonNthCalledWith negative throw matcher error for n that is not integer 1`] = `
"expect(received).not.sinonNthCalledWith(n, ...expected)

Matcher error: n must be a positive integer

n has type:  number
n has value: Infinity"
`;

exports[`sinonNthCalledWith positive throw matcher error for n that is not integer 1`] = `
"expect(received).sinonNthCalledWith(n, ...expected)

Matcher error: n must be a positive integer

n has type:  number
n has value: 0.1"
`;

exports[`sinonNthCalledWith positive throw matcher error for n that is not positive integer 1`] = `
"expect(received).sinonNthCalledWith(n, ...expected)

Matcher error: n must be a positive integer

n has type:  number
n has value: 0"
`;

exports[`sinonNthCalledWith works only on stubs or sinon.spy 1`] = `
"expect(received).sinonNthCalledWith(n, ...expected)

Matcher error: received value must be a stub or spy function

Received has type:  function
Received has value: [Function fn]"
`;

exports[`sinonNthCalledWith works when not called 1`] = `
"expect(sinon.spy()).sinonNthCalledWith(n, ...expected)

n: 1
Expected: \\"foo\\", \\"bar\\"

Number of calls: 0"
`;

exports[`sinonNthCalledWith works with Immutable.js objects 1`] = `
"expect(sinon.spy()).not.sinonNthCalledWith(n, ...expected)

n: 1
Expected: not Immutable.Map {\\"a\\": {\\"b\\": \\"c\\"}}, Immutable.Map {\\"a\\": {\\"b\\": \\"c\\"}}

Number of calls: 1"
`;

exports[`sinonNthCalledWith works with Map 1`] = `
"expect(sinon.spy()).not.sinonNthCalledWith(n, ...expected)

n: 1
Expected: not Map {1 => 2, 2 => 1}

Number of calls: 1"
`;

exports[`sinonNthCalledWith works with Map 2`] = `
"expect(sinon.spy()).sinonNthCalledWith(n, ...expected)

n: 1
- Expected
+ Received

  Map {
-   \\"a\\" => \\"b\\",
-   \\"b\\" => \\"a\\",
+   1 => 2,
+   2 => 1,
  },

Number of calls: 1"
`;

exports[`sinonNthCalledWith works with Set 1`] = `
"expect(sinon.spy()).not.sinonNthCalledWith(n, ...expected)

n: 1
Expected: not Set {1, 2}

Number of calls: 1"
`;

exports[`sinonNthCalledWith works with Set 2`] = `
"expect(sinon.spy()).sinonNthCalledWith(n, ...expected)

n: 1
- Expected
+ Received

  Set {
-   3,
-   4,
+   1,
+   2,
  },

Number of calls: 1"
`;

exports[`sinonNthCalledWith works with arguments that don't match 1`] = `
"expect(sinon.spy()).sinonNthCalledWith(n, ...expected)

n: 1
Expected: \\"foo\\", \\"bar\\"
Received: \\"foo\\", \\"bar1\\"

Number of calls: 1"
`;

exports[`sinonNthCalledWith works with arguments that don't match in number of arguments 1`] = `
"expect(sinon.spy()).sinonNthCalledWith(n, ...expected)

n: 1
Expected: \\"foo\\", \\"bar\\"
Received: \\"foo\\", \\"bar\\", \\"plop\\"

Number of calls: 1"
`;

exports[`sinonNthCalledWith works with arguments that don't match with matchers 1`] = `
"expect(sinon.spy()).sinonNthCalledWith(n, ...expected)

n: 1
Expected: Any<String>, Any<Number>
Received: \\"foo\\", \\"bar\\"

Number of calls: 1"
`;

exports[`sinonNthCalledWith works with arguments that don't match with matchers even when argument is undefined 1`] = `
"expect(sinon.spy()).sinonNthCalledWith(n, ...expected)

n: 1
Expected: \\"foo\\", Any<String>
Received: \\"foo\\", undefined

Number of calls: 1"
`;

exports[`sinonNthCalledWith works with arguments that match 1`] = `
"expect(sinon.spy()).not.sinonNthCalledWith(n, ...expected)

n: 1
Expected: not \\"foo\\", \\"bar\\"

Number of calls: 1"
`;

exports[`sinonNthCalledWith works with arguments that match with matchers 1`] = `
"expect(sinon.spy()).not.sinonNthCalledWith(n, ...expected)

n: 1
Expected: not Any<String>, Any<String>
Received:     0, [\\"foo\\", \\"bar\\"]

Number of calls: 1"
`;

exports[`sinonNthCalledWith works with three calls 1`] = `
"expect(sinon.spy()).not.sinonNthCalledWith(n, ...expected)

n: 1
Expected: not \\"foo1\\", \\"bar\\"
Received
->     1:     \\"foo1\\", \\"bar\\"
       2:     \\"foo\\", \\"bar1\\"

Number of calls: 3"
`;

exports[`sinonNthCalledWith works with trailing undefined arguments if requested by the match query 1`] = `
"expect(sinon.spy()).not.sinonNthCalledWith(n, ...expected)

n: 1
Expected: not \\"foo\\", undefined

Number of calls: 1"
`;

exports[`sinonNthCalledWith works with trailing undefined arguments when explicitly requested as optional by matcher 1`] = `
"expect(sinon.spy()).not.sinonNthCalledWith(n, ...expected)

n: 1
Expected: not \\"foo\\", optionalFn<>
Received:     0, [\\"foo\\", undefined]

Number of calls: 1"
`;
